File: LogicAndTrick.WikiCodeParser/Tags/QuickLinkTag.cs
using System.Collections.Generic;
using System.Text.RegularExpressions;
using LogicAndTrick.WikiCodeParser.Nodes;

namespace LogicAndTrick.WikiCodeParser.Tags
{
    public class QuickLinkTag : Tag
    {
        public QuickLinkTag()
        {
            Token = null;
            Element = "a";
            MainOption = "url";
            Options = new[] {"url"};
        }

        public override bool Matches(State state, string token)
        {
            var pt = state.PeekTo("]");
            if (string.IsNullOrEmpty(pt)) return false;

            pt = pt.Substring(1);
            return pt.Length > 0 && !pt.Contains("\n") && Regex.IsMatch(pt, @"^([a-z]{2,10}://[^\]]*?)(?:\|([^\]]*?))?", RegexOptions.IgnoreCase);
        }

        public override INode Parse(Parser parser, ParseData data, State state, string scope)
        {
            var index = state.Index;

            if (state.Next() != '[')
            {
                state.Seek(index, true);
                return null;
            }

            var str = state.ScanTo("]");
            if (state.Next() != ']')
            {
                state.Seek(index, true);
                return null;
            }

            var match = Regex.Match(str, @"^([a-z]{2,10}://[^\]]*?)(?:\|([^\]]*?))?$", RegexOptions.IgnoreCase);
            if (!match.Success)
            {
                state.Seek(index, true);
                return null;
            }

            var url = match.Groups[1].Value;
            var text = match.Groups[2].Length > 0 ? match.Groups[2].Value : url;
            var options = new Dictionary<string, string> {{"url", url}};
            if (!Validate(options, text))
            {
                state.Seek(index, true);
                return null;
            }

            url = System.Web.HttpUtility.HtmlAttributeEncode(url);
            var before = $"<{Element} href=\"{url}\">";
            var after = $"</{Element}>";

            var content = new PlainTextNode(text);
            return new HtmlNode(before, content, after)
            {
                PlainAfter = match.Groups[2].Length > 0 ? $" ({url})" : ""
            };
        }

        public override bool Validate(Dictionary<string, string> options, string text)
        {
            var url = text;
            if (options.ContainsKey("url")) url = options["url"];
            return !url.Contains("<script") && Regex.IsMatch(url, @"^([a-z]{2,10}://)?([^]""\n ]+?)", RegexOptions.IgnoreCase);
        }
    }
}

File: php/src/Tags/QuickLinkTag.php
<?php

namespace LogicAndTrick\WikiCodeParser\Tags;

use LogicAndTrick\WikiCodeParser\HtmlHelper;
use LogicAndTrick\WikiCodeParser\Nodes\HtmlNode;
use LogicAndTrick\WikiCodeParser\Nodes\INode;
use LogicAndTrick\WikiCodeParser\Nodes\UnprocessablePlainTextNode;
use LogicAndTrick\WikiCodeParser\ParseData;
use LogicAndTrick\WikiCodeParser\Parser;
use LogicAndTrick\WikiCodeParser\State;
use LogicAndTrick\WikiCodeParser\TagParseContext;

class QuickLinkTag extends Tag
{
    public function __construct()
    {
        parent::__construct();
        $this->token = null;
        $this->element = 'a';
        $this->mainOption = 'url';
        $this->options = ['url'];
    }

    public function Matches(State $state, ?string $token, TagParseContext $context): bool
    {
        $pt = $state->PeekTo(']');
        if (!$pt || $pt == '') return false;

        $pt = substr($pt, 1);
        return strlen($pt) > 0 && !str_contains($pt, "\n") && preg_match('/^([a-z]{2,10}:\/\/[^\]]*?)(?:\|([^\]]*?))?/i', $pt);
    }

    public function Parse(Parser $parser, ParseData $data, State $state, string $scope, TagParseContext $context): INode|null
    {
        $index = $state->index;

        if ($state->Next() != '[') {
            $state->Seek($index, true);
            return null;
        }

        $str = $state->ScanTo(']');
        if ($state->Next() != ']') {
            $state->Seek($index, true);
            return null;
        }

        $success = preg_match('/^([a-z]{2,10}:\/\/[^\]]*?)(?:\|([^\]]*?))?$/i', $str, $match);
        if (!$success) {
            $state->Seek($index, true);
            return null;
        }

        $url = $match[1];
        $text = isset($match[2]) && strlen($match[2]) > 0 ? $match[2] : $url;
        $options = [ 'url' => $url ];
        if (!$this->Validate($options, $text)) {
            $state->Seek($index, true);
            return null;
        }

        $url = HtmlHelper::AttributeEncode($url);
        $before = "<$this->element href=\"$url\">";
        $after = "</$this->element>";

        $content = new UnprocessablePlainTextNode($text);
        $ret = new HtmlNode($before, $content, $after);
        $ret->plainAfter = isset($match[2]) && strlen($match[2]) > 0 ? " ($url)" : '';
        return $ret;
    }

    public function Validate(array $options, string $text): bool
    {
        $url = $text;
        if (isset($options['url'])) $url = $options['url'];
        return !str_contains($url, '<script') && preg_match('/^([a-z]{2,10}:\/\/)?([^\]"\n ]+?)$/i', $url);
    }
}